src/ws/cockpit-session-client-certificate.c

```diff
@@ -43,65 +43,50 @@
-                             "cockpit-wsinstance-https@([0-9a-f]{64}).service$"
-#define CGROUP_REGEX_FLAGS   (REG_EXTENDED | REG_NEWLINE)
-#define CGROUP_REGEX_GROUPS  3   /* number of groups, including the complete match */
-#define CGROUP_REGEX_MATCH   2   /* the group which contains the instance */
-
-/* get our cgroup, map it to a systemd unit instance name
- * looks like 0::/system.slice/system-cockpit\x2dwsinstance\x2dhttps.slice/cockpit-wsinstance-https@123abc.service
- * returns "123abc" instance name (static string)
+/* Reads the cgroupsv2-style /proc/[pid]/cgroup file of the process,
+ * including "0://" prefix and newline.
+ *
+ * In case of cgroupsv1, look for the name=systemd controller, and fake
+ * it.
  */
-static const char *
-get_ws_https_instance (void)
+static char *
+read_proc_self_cgroup (size_t *out_length)
 {
-  int r;
-  static char buf[1024];
-  regmatch_t pmatch[CGROUP_REGEX_GROUPS];
-  regex_t preg;
-  int fd;
-
-  /* read /proc/self/cgroup */
-  fd = open ("/proc/self/cgroup", O_RDONLY | O_CLOEXEC | O_NOFOLLOW);
-  if (fd < 0)
-    {
-      warn ("Failed to open /proc/self/cgroup");
-      return NULL;
-    }
-
-  do
-    r = read (fd, buf, sizeof buf);
-  while (r < 0 && errno == EINTR);
-  if (r < 0)
-    {
-      warn ("Failed to read /proc/self/cgroup");
-      close (fd);
-      return NULL;
-    }
-  close (fd);
-  if (r == 0 || r >= sizeof buf)
-    {
-      warnx ("Read invalid size %i from /proc/self/cgroup", r);
-      return NULL;
-    }
-  buf[r] = '\0';
+  FILE *fp = fopen ("/proc/self/cgroup", "r");
 
-  /* extract the instance name */
-  r = regcomp (&preg, CGROUP_REGEX, CGROUP_REGEX_FLAGS);
-  assert (r == 0);
+  if (fp == NULL)
+    return NULL;
 
-  r = regexec (&preg, buf, CGROUP_REGEX_GROUPS, pmatch, 0);
-  regfree (&preg);
-  if (r != 0)
+  /* Support cgroups v1 by looping.
+   * Once we no longer need this support, we can drop the loop, switch
+   * to fread(), and just return the entire content of the file.
+   */
+  char buffer[1024];
+  char *result = NULL;
+  while (fgets (buffer, sizeof buffer, fp))
     {
-      warnx ("Not running in a template cgroup, unable to parse systemd unit instance.\n\n/proc/self/cgroups content follows:\n%s\n", buf);
-      return NULL;
+      assert (strchr (buffer, '\n')); /* ...or else, kernel bug. */
+
+      if (strncmp (buffer, "0::", 3) == 0)
+        {
+          /* cgroupsv2 (or hybrid) case.  Return the entire line. */
+          result = strdupx (buffer);
+          break;
+        }
+      else if (strncmp (buffer, "1:name=systemd:", 15) == 0)
+        {
+          /* cgroupsv1.  Rewrite to what we'd expect from cgroupsv2. */
+          asprintfx (&result, "0::/%s", buffer + 15);
+          break;
+        }
     }
 
-  buf[pmatch[CGROUP_REGEX_MATCH].rm_eo] = '\0';
+  fclose (fp);
 
-  return buf + pmatch[CGROUP_REGEX_MATCH].rm_so;
+  if (out_length)
+    *out_length = strlen (result);
```

Comment:
This will crash if `result` is NULL, i.e. if scanning the cgroup file had no match. This Should Not Happen™, but there has been wild stuff out there, like old LXC emulating /proc and /sys for unprivileged containers in the past, or maybe even custom systemd configuration. Better add an `&& result` here.

Reply:
added as an assert.  I don't want to deal with this function returning NULL.